examples/ecs/nondeterministic_system_order.rs

```diff
@@ -0,0 +1,83 @@
+//! By default, Bevy systems run in parallel with each other.
+//! Unless the order is explicitly specified, their relative order is nondeterministic.
+//!
+//! In many cases, this doesn't matter and is in fact desirable!
+//! Consider two systems, one which writes to resource A, and the other which writes to resource B.
+//! By allowing their order to be arbitrary, we can evaluate them greedily, based on the data that is free.
+//! Because their data accesses are **compatible**, there is no **observable** difference created based on the order they are run.
+//!
+//! But if instead we have two systems mutating the same data, or one reading it and the other mutating,
+//! than the actual observed value will vary based on the nondeterministic order of evaluation.
+//! These observable conflicts are called **system execution order ambiguities**.
+//!
+//! This example demonstrates how you might detect and resolve (or silence) these ambiguities.
+
+use bevy::{ecs::schedule::ReportExecutionOrderAmbiguities, prelude::*};
+
+fn main() {
+    App::new()
+        // This resource controls the reporting strategy for system execution order ambiguities
+        .insert_resource(ReportExecutionOrderAmbiguities)
+        .init_resource::<A>()
+        .init_resource::<B>()
+        // This pair of systems has an ambiguous order,
+        // as their data access conflicts, and there's no order between them.
+        .add_system(reads_a)
+        .add_system(writes_a)
+        // This trio of systems has conflicting data access,
+        // but it's resolved with an explicit ordering
+        .add_system(adds_one_to_b)
+        .add_system(doubles_b.after(adds_one_to_b))
+        // This system isn't ambiguous with adds_one_to_b,
+        // due to the transitive ordering created.
+        .add_system(reads_b.after(doubles_b))
+        // This system will conflict with all of our writing systems
+        // but we've silenced its ambiguity with adds_one_to_b.
+        // This should only be done in the case of clear false positives:
+        // leave a comment in your code justifying the decision!
+        .add_system(reads_a_and_b.ambiguous_with(adds_one_to_b))
+        // Be mindful, internal ambiguities are reported too!
+        // If there are any ambiguities due solely to DefaultPlugins,
+        // or between DefaultPlugins and any of your third party plugins,
+        // please file a bug with the repo responsible!
+        // Only *you* can prevent nondeterministic bugs due to greedy parallelism.
+        .add_plugins(DefaultPlugins)
+        // We're only going to run one frame of this app,
+        // to make sure we can see the warnings at the start.
+        .update();
```

Comment:
Perhaps instead of the one `.update()` add a system that sends an `AppExit` Event, which must run after all other systems are finished.
This can be used to explain, that `ReportExecutionOrderAmbiguities` is only concerned with the declared access of `SystemParam`s, not other forms of dependencies, that cannot be tracked/expressed in the Type system.

Reply:
Though thinking further, due to the implementation of exiting the app, it `.after(...)` is actually unnecessary here, as the actual Shutdown happens at the end of the current frame.

Are there any other practical Patterns that might have a conceptional Dependency, that can't be encoded in types (And do not use Interior Mutability)?

Reply:
I decided to borrow this AppExit idea, but I think we should keep this example simple. There's already a lot to learn here!